component/main.jsonnet

```
@@ -19,5 +22,22 @@ local rules = import 'rules.jsonnet';
   '01_prometheus': prometheus,
   '02_alertmanager': alertmanager,
   '10_federation': federation,
-  'prometheus-rules': rules,
+  'rules/00_kube-prometheus': rules,
+} + {
+  ['rules/' + group_name]: prom.PrometheusRule(group_name) {
+    metadata+: {
+      namespace: params.namespace,
+      labels+: {
+        prometheus: params.prometheusInstance,
+        role: 'alert-rules',
+      },
+    },
+    spec+: {
+      groups+: [ {
+        name: group_name,
+        rules: params.rules[group_name],
+      } ],
+    },
+  }
+  for group_name in std.objectFields(params.rules)
```

The current implementation doesn't allow users to patch or remove rules which were added higher up in the hierarchy because it's impossible to make changes to existing list entries in reclass.

You could instead consider to do something like the following (this would still support both recording and alerting rules) to allow users to patch and remove rules (and whole groups) in the hierarchy:

```suggestion
    spec+: {
      groups+: [ {
        name: group_name,
        rules: [
          local rnamekey = std.splitLimit(rname, ':', 1);
          params.rules[group_name][rname] {
            [rnamekey[0]]: rnamekey[1],
          }
          for rname in std.objectFields(params.rules[group_name])
          if params.rules[group_name][rname] != null
        ],
      } ],
    },
  }
  for group_name in std.objectFields(params.rules)
  if params.rules[group_name] != null
```

---

I'm not sure If we really need this right now.
We are starting with the redesign of the whole monitoring stack soon anyways, where this will be implemented anyways. Do you think it makes sense to add this here now?

---

I don't think it hurts and the proposed change is in line with the best practices for components.

Additionally, people will start using the feature once it's available, so imho it makes sense to implement it in a way that allows doing the things that are generally possible, such as patching/removing configs in the hierarchy.